### DISCUSSION TITLE
Support for scipy.sparse.linalg.LinearOperator

### DISCUSSION THREAD

**Discussion by stephane-caron** [qpsolvers/qpsolvers#172] [Ideas]
From @wertersilva in https://github.com/stephane-caron/qpsolvers/discussions/147#discussioncomment-4883908:

> Does any of the solvers support scipy.sparse.linalg.LinearOperator? Thanks.

**Comment by stephane-caron** [comment]
Looking at the docs for [`LinearOperator`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html) I don't see exactly what you mean. The example from the docs is:

```python
>>> import numpy as np
>>> from scipy.sparse.linalg import LinearOperator
>>> def mv(v):
...    return np.array([2*v[0], 3*v[1]])
>>> A = LinearOperator((2,2), matvec=mv)
>>> A
<2x2 _CustomLinearOperator with dtype=float64>
>>> A.matvec(np.ones(2))
array([ 2.,  3.])
>>> A * np.ones(2)
array([ 2.,  3.])
```

Can you provide a similar example for the usage you'd like?

**Comment by wertersilva** [comment]
The website https://pylops.readthedocs.io/en/stable/ better illustrates the use of LinearOperator. We can subclass it to create linear operators that act the same way as matrices by implementing the methods _matcvec and _rmatvec. This allows us to encapsulate complex matrix-free operators such as those representing wave equation stencils and still use A*x-like matrix-vector multiplication notation to describe the action of those operators. 

For example, the signature of solve_qp below seems to accept numpy arrays and scipy sparse matrices for the P argument,  but not LinearOperators, like those defined in the mentioned webpage, which would be great.

Signature:
solve_qp(
    P: Union[numpy.ndarray, scipy.sparse._csc.csc_matrix],
    q: numpy.ndarray,
    G: Union[numpy.ndarray, scipy.sparse._csc.csc_matrix, NoneType] = None,
    h: Optional[numpy.ndarray] = None,
    A: Union[numpy.ndarray, scipy.sparse._csc.csc_matrix, NoneType] = None,
    b: Optional[numpy.ndarray] = None,
    lb: Optional[numpy.ndarray] = None,
    ub: Optional[numpy.ndarray] = None,
    solver: Optional[str] = None,
    initvals: Optional[numpy.ndarray] = None,
    sym_proj: bool = False,
    verbose: bool = False,
    **kwargs,
) -> Optional[numpy.ndarray]

**Reply by stephane-caron** [reply]
OK, so if I follow you correctly, one usage example would look like:

```python
Pop = pylops.MatrixMult(P)  # this is one way to build a LinearOperator
x = solve_qp(Pop, q, G, h, A, b, solver="...")
```

As far as I know, none of the currently available solvers is ready to handle ``LinearOperator``s. That's because quadratic programming is defined [for real-valued matrices](https://en.wikipedia.org/wiki/Quadratic_programming#Problem_formulation). If what you have at hand is a ``LinearOperator``, you can always map it back to a matrix using its [``matmat``](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.matmat.html#scipy.sparse.linalg.LinearOperator.matmat) function:

```python
P = Pop.matmat(np.eye(n))
x = solve_qp(P, q, G, h, A, b, solver="...")
```
